src/main/scala/cognite/spark/v1/CdpConnector.scala

```diff
@@ -20,19 +22,18 @@ case class Login(user: String, loggedIn: Boolean, project: String, projectId: Lo
 object CdpConnector {
   @transient lazy val cdpConnectorExecutionContext: ExecutionContext =
     ExecutionContext.fromExecutor(
-      new ThreadPoolExecutor(
-        0,
-        Constants.MaxConcurrentRequests,
-        60L,
-        TimeUnit.SECONDS,
-        new SynchronousQueue()))
+      Executors.newFixedThreadPool(
+        Math.max(Runtime.getRuntime().availableProcessors(), 4) * 2,
+        new ThreadFactoryBuilder().setDaemon(true).setNameFormat("CDF-Spark-Datasource-%d").build()
```

Comment:
`background threads = daemon threads` in JVM world. We either need to make an interface for shutting down our library, or add a shutdown hook, or turn the threads into background threads, to prevent our thread pool from hanging the host application. In our case it's not critical if the thread pool dies abrubtly during exit, and it's safer than handling it ourself in a shutdown hook, where we still may end up locking the JVM. 

In general it's safe to set threads as daemon, and it's a fair assumption that a user of a library doesn't exit from the main thread (foreground thread) before usage of a library is done, in which case the user risk undefined behaviour. 

Foreground threads are only useful in cases where one run critical logic, and you would like the operation to finish before the application exits, however that would probably be solved in other ways eg. through some message pump and is also tightly coupled with application logic and not library logic.

Reply:
We may also go for the shutdown hook approach, however we may end up hanging the application as long as there are writes batches queued, which may not be desired behaviour. Also there is a risk for us failing to make a stable shutdown hook (eg. an exception happens, which stops us from shutting down things).

Reply:
I'm fine with daemon threads, just wanted to have a better understanding of why, and your explanation is a good one. I noted they use it here as well: https://github.com/snowflakedb/spark-snowflake/blob/70713d339b887ae7c57d22d1ee10d58a3a83303e/src/main/scala/net/snowflake/spark/snowflake/SnowflakeJDBCWrapper.scala#L65

If they're using daemon threads in their data source, we should probably do it in ours as well.

But again, while the reasons are good and well explained by you here in GH, we should have comments in the code explaining this as well.

Reply:
Yes, I don't disagree with explaining it in a comment.

Reply:
Also, it's not only the datasources using it, but Spark itself set all their threads to daemon, regardless of them also having a shutdown hook. I remember from my .Net times, this was also an issue there (Thread API was more or less copied from Java when .Net was designed), however I think the default foreground behaviour is unexpected and undesired, however the default has sticked for back compatibility I believe. If the Thread API was designed today I think they would make the Threads background/daemon by default, as this is something you instead want to control by using semaphores and such, or some sort of concurrent queue for messaging to stop a thread safely (or quit after a timeout).